keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1603 @@
+address.) Furthermore, because of how Kubernetes Services evolved in
+parallel with the existing kube-proxy implementation, we have ended up
+with a dual-stack Service semantics that is most easily implemented by
+handling IPv4 and IPv6 completely separately anyway.)
+
+#### Communicating with the kernel nftables subsystem
+
+We will use the `nft` command-line tool to read and write rules, much
+like how we use command-line tools in the `iptables` and `ipvs`
+backends.
+
+However, the `nft` tool is mostly just a thin wrapper around
+`libnftables`, so any golang API that wraps the `nft` command-line
+could easily be rewritten to use `libnftables` directly (via a cgo
+wrapper) in the future if that seemed like a better idea. (In theory
+we could also use netlink directly, without needing cgo or external
+libraries, but this would probably be a bad idea; `libnftables`
+implements quite a bit of functionality on top of the raw netlink
+API.)
+
+When reading data from the kernel (`nft list ...`), `nft` outputs the
+data in a nested "object" form:
+
+```
+table ip kube_proxy {
+  comment "Kubernetes service proxying rules";
+
+  chain services {
+    ip daddr . ip protocol . th dport vmap @service_ips
+  }
+}
+```
+
+(This is the "native" nftables syntax, but the tools also support a
+JSON syntax that may be easier for us to work with...)
+
+When writing data to the kernel, `nft` accepts the data in either the
+same "object" form used by `nft list`, or in the form of a set of
+`nft` command lines without the leading "`nft`" (which are then
+executed atomically):
+
+```
+add table ip kube_proxy { comment "Kubernetes service proxying rules"; }
+add chain ip kube_proxy services
+add rule ip kube_proxy services ip daddr . ip protocol . th dport vmap @service_ips
+```
+
+The "object" form is more logical and easy to understand, but the
+"command" form is better for dynamic usage. In particular, it allows
+you to add and remove individual chains, rules, map/set elements, etc,
+without needing to also include the chains/rules/elements that you are
+not modifying.
+
+The examples below all show the "object" form of data, but it should
+be understood that these are examples of what would be seen in `nft
+list` output after kube-proxy creates the rules (with additional
+`#`-preceded comments added to help the KEP reader), not examples of
+the data we will actually be passing to `nft`.
+
+The examples below are also all IPv4-specific, for simplicity. When
+actually writing out rules for nft, we will need to switch between,
+e.g., "`ip daddr`" and "`ip6 daddr`" appropriately, to match an IPv4
+or IPv6 destination address. This will actually be fairly simple
+because the `nft` command lets you create "variables" (really
+constants) and substitute their values into the rules. Thus, we can
+just always have the rule-generating code write "`$IP daddr`", and
+then pass either "`-D IP=ip`" or "`-D IP=ip6`" to `nft` to fix it up.)
```

Comment:
I prefer remove as much as possible any dependency on the userspace tool `nft` and do all this in golang ... is easier to understand and removes any chance of hitting bugs on these features

Reply:
So, as I've been working on this, I've discovered that the situation with the nftables API is kind of more ugly than I'd thought. The underlying netlink API (which is not really documented anywhere) is very very different from the text-based command-line API (which is documented in the man page, the wiki, lots of random third-party blog posts, etc). There is also a JSON API, which is mostly a mapping of the netlink API (meaning that it is also poorly documented and hard to translate to and from the textual command-line API).

So we have to choose between:
1. use the API that 99% of documentation and examples on the web use
2. use the API that is closer to the bits and less prone to hitting random bugs

I feel like for maintainability purposes, we really have to go with 1.

Reply:
I was referring more to the use of "special" feature as variable replacement ... I think that 1. is ok if we achieve the same degree that we got in iptables